src/IterableCodeExtractor.php

```diff
@@ -90,61 +90,83 @@ public static function fromDirectory( $dir, Translations $translations, array $o
 		}
 
-		if ( in_array( $file->getBasename(), $include, true ) ) {
-			return true;
+		if ( in_array( $file->getBasename(), $matchers, true ) ) {
+			return 10;
 		}
 
 		// Check for more complex paths, e.g. /some/sub/folder.
 		$root_relative_path = str_replace( static::$dir, '', $file->getPathname() );
-		foreach ( $include as $path_or_file ) {
+
+		foreach ( $matchers as $path_or_file ) {
 			$pattern = preg_quote( str_replace( '*', '__wildcard__', $path_or_file ) );
-			$pattern = '/' . str_replace( '__wildcard__', '(.+)', $pattern );
+			$pattern = '(^|/)' . str_replace( '__wildcard__', '(.+)', $pattern );
+
+			$base_score = count(
+				array_filter(
+					explode( '/', $path_or_file ),
+					function ( $component) { return $component !== '*'; }
+				)
+			);
+			if ( 0 === $base_score ) {
+				// If the matcher is simply * it gets a score above the implicit score but below 1.
+				$base_score = 0.2;
+			}
+
 
 			if (
-				false !== mb_ereg( $pattern, $root_relative_path . '$' ) &&
-				false !== mb_ereg( $pattern, $root_relative_path . '/' )
+				false === strpos( $path_or_file, '*' ) &&
+				false !== mb_ereg( $pattern . '$', $root_relative_path )
 			) {
-				return true;
+				return $base_score * 10;
+			}
+
+			if ( false !== mb_ereg( $pattern . '(/|$)', $root_relative_path ) ) {
+				return $base_score;
 			}
 		}
 
-		return false;
+		return 0;
 	}
 
 	/**
-	 * Determines whether a file is valid based on the exclude option.
+	 * Determines whether or not a directory has children that may be matched.
 	 *
-	 * @param SplFileInfo $file    File or directory.
-	 * @param array       $exclude List of files and directories to skip.
-	 * @return bool
+	 * @param SplFileInfo $dir      Directory.
+	 * @param array       $matchers List of files and directories to match.
+	 * @return bool Whether or not there are any matchers for children of this directory.
 	 */
-	protected static function isExcluded( SplFileInfo $file, array $exclude = [] ) {
-		if ( empty( $exclude ) ) {
+	protected static function containsMatchingChildren( SplFileInfo $dir, array $matchers = [] ) {
+		if ( empty( $matchers ) ) {
 			return false;
 		}
 
-		if ( in_array( $file->getBasename(), $exclude, true ) ) {
-			return true;
-		}
+		$root_relative_path = str_replace( static::$dir, '', $dir->getPathname() );
 
-		// Check for more complex paths, e.g. /some/sub/folder.
-		foreach ( $exclude as $path_or_file ) {
-			$pattern = preg_quote( str_replace( '*', '__wildcard__', $path_or_file ) );
-			$pattern = '/' . str_replace( '__wildcard__', '(.+)', $pattern ) . '$';
+		foreach ( $matchers as $path_or_file ) {
+			if (
+				false === strpos( $path_or_file, '*' ) &&
+				0 === strpos( $path_or_file . '/', $root_relative_path )
+			) {
+				return true;
+			}
+
+			$base = current( explode( '*', $path_or_file ) );
 
-			$root_relative_path = str_replace( static::$dir, '', $file->getPathname() );
-			if ( false !== mb_ereg( $pattern, $root_relative_path ) ) {
+			if (
```

Comment:
This needs a comment as well. I'm not 100% sure what the reason is for checking both ways, I assume this is done for being overly conservative. This certainly needs an explanation.

Reply:
Needed for partial wildcards like this:

Given an include wildcard of: `/dir/wp-*` any directory starting with `/dir/wp-` should always be checked.

Whereas for normal wildcards the check is the other way around: `one/two/three/*.js` means that a directory called `one` should be checked to see if it starts with the matcher.